src/server/pps/server/pipeline_controller.go

```diff
@@ -188,48 +190,34 @@ func (pc *pipelineController) step(timestamp time.Time) (retErr error) {
 		tracing.TagAnySpan(span, "pollpipelines-event", "true")
 	}
 	defer func() {
-		tracing.FinishAnySpan(span, "err", retErr)
+		tracing.FinishAnySpan(span, "err", finalErr)
 	}()
 
-	// query pipelineInfo
-	var pi *pps.PipelineInfo
-	var err error
-	if pi, err = pc.tryLoadLatestPipelineInfo(); err != nil && collection.IsErrNotFound(err) {
-		// if the pipeline info is not found, interpret the operation as a delete
+	// if we fail to create a new step, there was an error querying the pipeline info, and there's nothing we can do
+	step, err := pc.newStep()
+	if err != nil {
+		finalErr = err
+		log.Errorf("PPS master: failed to set up step data to handle event for pipeline '%s': %v", pc.pipeline,
+			errors.Wrapf(err, "failing pipeline %q", pc.pipeline))
+	}
+
+	// interpret the event as a delete operation
+	if step == nil {
 		if err := pc.deletePipelineResources(); err != nil {
+			finalErr = err
 			log.Errorf("PPS master: error deleting pipelineController resources for pipeline '%s': %v", pc.pipeline,
 				errors.Wrapf(err, "failing pipeline %q", pc.pipeline))
 		}
-		pc.tryFinish()
-		return nil
-	} else if err != nil {
-		return err
-	}
-
-	tracing.TagAnySpan(pc.ctx,
-		"current-state", pi.State.String(),
-		"spec-commit", pretty.CompactPrintCommitSafe(pi.SpecCommit))
-	// add pipeline auth
-	// the pipelineController's context is authorized as pps master, but we want to switch to the pipeline itself
-	// first clear the cached WhoAmI result from the context
-	pachClient := pc.env.GetPachClient(middleware_auth.ClearWhoAmI(pc.ctx))
-	pachClient.SetAuthToken(pi.AuthToken)
-	stepCtx := pachClient.Ctx()
-
-	step := &pcStep{
-		pc:           pc,
-		ctx:          stepCtx,
-		pipelineInfo: pi,
+		return true
 	}
 
-	// Process the pipeline event
-	var errCount int
 	var stepErr stepError
+	errCount := 0
 	err = backoff.RetryNotify(func() error {
 		// set pc.rc
 		// TODO(msteffen) should this fail the pipeline? (currently getRC will restart
 		// the pipeline indefinitely)
-		if err := step.getRC(stepCtx, noExpectation); err != nil && !errors.Is(err, errRCNotFound) {
+		if err := step.getRC(noExpectation); err != nil && !errors.Is(err, errRCNotFound) {
```

Comment:
minor style nit, but now that I see the code laid out, with `getRC` and `run()` together in this retry loop, I wonder if it would make sense to rename `run()` to `tryUpdateRC()` or some such—make the retry loop the "RC portion" of the logic, basically.

Reply:
yeah that's much more descriptive than run. Do you have a preference between `tryUpdateRC()` and `updateRcAndState()`?